aurora/contracts/src/AuroraErc20FastBridge.sol

```diff
@@ -281,39 +286,46 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
         emit WithdrawFromNear(tokenId, amount);
     }
 
-    function withdraw(string memory token) public whenNotPaused {
+    function withdraw(string calldata token) external whenNotPaused {
         uint128 signerBalance = balance[token][msg.sender];
 
         require(signerBalance > 0, "The signer token balance = 0");
+        require(near.wNEAR.balanceOf(address(this)) > 0, "Not enough wNEAR balance of AuroraErc20FastBridge");
 
         bytes memory args = bytes(
             string.concat(
-                '{"receiver_id":',
+                '{"receiver_id": "',
                 auroraEngineAccountIdOnNear,
-                '"amount": "',
+                '", "amount": "',
                 Strings.toString(signerBalance),
                 '", "msg": "',
                 _addressToString(msg.sender),
                 '"}'
             )
         );
+        balance[token][msg.sender] -= signerBalance;
+
         PromiseCreateArgs memory callWithdraw = near.call(token, "ft_transfer_call", args, 1, WITHDRAW_NEAR_GAS);
         bytes memory callbackArg = abi.encodeWithSelector(this.withdrawCallback.selector, msg.sender, token);
         PromiseCreateArgs memory callback = near.auroraCall(address(this), callbackArg, 0, BASE_NEAR_GAS);
 
         callWithdraw.then(callback).transact();
     }
 
-    function withdrawCallback(address signer, string memory token) public onlyRole(CALLBACK_ROLE) {
+    function withdrawCallback(address signer, string calldata token, uint128 amount) external onlyRole(CALLBACK_ROLE) {
         require(
             AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
             "ERROR: The `Withdraw` XCC is fail"
         );
 
         uint128 transferredAmount = _stringToUint(AuroraSdk.promiseResult(0).output);
+        uint128 refundAmount = amount - transferredAmount;
+        
+        if (refundAmount > 0) {
+            balance[token][signer] += refundAmount;
```

Comment:
But what if the `XXC` fails, for example, because we didn't call `withdrawFromNear` before calling `withdraw`?
Are the user's tokens stuck forever?

Reply:
fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/96/commits/57a45daffb0c7bd2c0e09f8ba131ca4c1826c133